### PR TITLE
Move substreams processing to block stream

### PR DESCRIPTION
**Pull request by mangas**
Improves perfromance based on [this issue](https://github.com/graphprotocol/graph-node/issues/4831)

### CODE REVIEW

File: chain/substreams/src/trigger.rs [hunk 1]
```diff
@@ -189,95 +183,37 @@ where
         _subgraph_metrics: &Arc<graph::prelude::SubgraphInstanceMetrics>,
         _instrument: bool,
     ) -> Result<BlockState<Chain>, MappingError> {
-        for entity_change in block.changes.entity_changes.iter() {
-            match entity_change.operation() {
-                Operation::Unset => {
+        for parsed_change in block.parsed_changes.clone().into_iter() {
```
Comment (maoueh) [inline chain/substreams/src/trigger.rs:186]: This `clone` seems at first sight superflous. In each arm of the match, you either use a reference or clone, I didn't see any move in there.

Comment (mangas) [reply chain/substreams/src/trigger.rs:186]: I will check but I think this is because block is a &Arc so you can't move it out of there to call into_iter

Comment (maoueh) [reply chain/substreams/src/trigger.rs:186]: Seems that `state.entity_cache.remove(entity_key);` is the only move. I think cloning just this one would be better. I leave to you the decision.

Comment (maoueh) [reply chain/substreams/src/trigger.rs:186]: Scratch all my comment above, I just saw `state.entity_cache.set(key, entity)?;` which is a full move, I missed it because it was part of non-modified code. Sorry :)

Comment (mangas) [reply chain/substreams/src/trigger.rs:186]: Yeah that's it I was just re-checking and the set requires the move so we might as well clone it all ahead of time. No problem at all, thanks for checking!

File: chain/substreams/src/mapper.rs [hunk 2]
```diff
@@ -1,13 +1,27 @@
...
+use graph::data_source::CausalityRegion;
+use graph::prelude::BigDecimal;
+use graph::prelude::{async_trait, BigInt, BlockHash, BlockNumber, BlockPtr, Logger, Value};
+use graph::schema::InputSchema;
 use graph::substreams::Clock;
 use graph::substreams_rpc::response::Message as SubstreamsMessage;
 use prost::Message;
 
-pub struct Mapper {}
+pub struct Mapper {
```
Comment (lutter) [inline chain/substreams/src/mapper.rs:22]: It would be good to explain here what this is for, and how setting `schema` to `None` influences things

File: chain/substreams/src/mapper.rs [hunk 3]
```diff
@@ -89,3 +113,237 @@ impl SubstreamsMapper<Chain> for Mapper {
...
+                        .or_insert(Value::Null) = value;
+                }
+                let entity = schema
+                    .make_entity(parsed_data)
+                    .map_err(anyhow::Error::from)?;
+
+                ParsedChanges::Upsert { key, entity }
+            }
+            entity_change::Operation::Delete => ParsedChanges::Delete(key),
+            entity_change::Operation::Unset => ParsedChanges::Unset,
```
Comment (lutter) [inline chain/substreams/src/mapper.rs:176]: I am not familiar with substreams vocab, but what does `Unset` mean in this context? Do we even need to keep that in the parsed changes? Might be good to add that as a comment to `ParsedChanges`

File: chain/substreams/src/chain.rs [hunk 4]
```diff
@@ -20,13 +19,22 @@ use graph::{
...
     slog::Logger,
 };
+
 use std::sync::Arc;
 
+#[derive(Debug, Clone)]
+pub enum ParsedChanges {
+    Unset,
+    Delete(EntityKey),
+    Upsert { key: EntityKey, entity: Entity },
```
Comment (lutter) [inline chain/substreams/src/chain.rs:29]: No more intermediate `HashMap` .. yay!

File: core/src/subgraph/runner.rs [hunk 5]
```diff
@@ -723,7 +725,7 @@ where
         for trigger in triggers {
             // Using an `EmptyStore` and clearing the cache for each trigger is a makeshift way to
             // get causality region isolation.
-            let schema = self.inputs.store.input_schema();
+            let schema = ReadStore::input_schema(&self.inputs.store);
```
Comment (lutter) [inline core/src/subgraph/runner.rs:728]: Why was that change necessary?

**Review by lutter — APPROVED** [review]

File: core/src/subgraph/runner.rs (hunk 5 above)
Comment (mangas) [reply core/src/subgraph/runner.rs:728]: I added the `input_schema` function to another trait which caused this call to be ambiguous.

File: chain/substreams/src/mapper.rs (hunk 3 above)
Comment (mangas) [reply chain/substreams/src/mapper.rs:176]: will do, Unset is just to preserve the code logic that's applied to protobuf. Unset there is the default value when a new unsupported value is used. Say you have an enum with 3 values,  if the server has a 4th value then when parsing it will default to Unset, works as a sentinel value of sorts. 

In this case, at parsing we don't "care" about what operation it is, we're just transforming so there is a 1:1 mapping of supported operations. Once we run the trigger processing then there will be an error if the operation is `Unset` as that's not a valid operation. I didn't want to change the responsibility of the block stream so this was the "easiest" way to maintain it

File: chain/substreams/src/mapper.rs (hunk 2 above)
Comment (mangas) [reply chain/substreams/src/mapper.rs:22]: will do!

File: chain/substreams/src/mapper.rs (hunk 3 above)
Comment (mangas) [reply chain/substreams/src/mapper.rs:176]: I will also add a comment here to make it clearer
